piptools/resolver.py

```diff
@@ -119,9 +119,9 @@ def resolve(self, max_rounds=10):
 
     def _check_constraints(self):
         for constraint in chain(self.our_constraints, self.their_constraints):
-            if constraint.link is not None and not constraint.editable:
-                msg = ('pip-compile does not support URLs as packages, unless they are editable. '
-                       'Perhaps add -e option?')
+            if ((constraint.link and not constraint.editable
+                 and not constraint.link.is_artifact and not is_pinned_requirement(constraint))):
```

Comment:
I'm having a hard time understanding this condition. Why is the `is_pinned_requirement()` in there?